### PR TITLE
feat(telemetry): Enhance telemetry

### LINKED ISSUE #10290: feat(telemetry): Report population capture and environment flag
- Start gathering instance patient and user population. I may break out users by their role, patient by portal or not etc.
- Add an environment variable to existing startup test to disable telemetry if a session flag is set in main.php. I currently use the session flag to disable in test because I could not find a way to set an environment flag that was in scope from main.php on test run. @kojiromike do you want to recommend the variable name for me to capture in telemetry init?

### PR DESCRIPTION
**Pull request by sjpadgett**
Add environment variable check to disable telemetry

Fixes #10290

### CODE REVIEW

File: interface/main/tabs/main.php [hunk 1]
```diff
@@ -41,6 +41,18 @@
 $allowRegisterDialog = $product_row['allowRegisterDialog'] ?? 0;
 $allowTelemetry = $product_row['allowTelemetry'] ?? null; // for dialog
 $allowEmail = $product_row['allowEmail'] ?? null; // for dialog
+
+// Check if telemetry is disabled via environment variable
+$env_name = 'OPENEMR_DISABLE_TELEMETRY';
```
Comment (kojiromike) [inline interface/main/tabs/main.php]: If you like, this variable could either disable telemetry, or if a non-empty string (email address?), it could auto-register an openemr instance. I might also suggest this could be a const.

File: interface/main/tabs/main.php [hunk 2]
```diff
@@ -41,6 +41,18 @@
...
+$val = getenv($env_name);
+if ($val === false || $val === '') {
+    $val = $_ENV[$env_name] ?? $_SERVER[$env_name] ?? null;
+}
+$disableTelemetry = $val !== null && in_array(strtolower(trim((string)$val)), ['1', 'true', 'yes', 'on', 'y'], true);
```
Comment (kojiromike) [inline interface/main/tabs/main.php]:
```suggestion
$disableTelemetry = filter_var($val, FILTER_VALIDATE_BOOLEAN);
```

will replace most of this – unless recognizing 'y' is very important, in which case

```
$disableTelemetry = $val === 'y' || filter_var($val, FILTER_VALIDATE_BOOLEAN);
```

Comment (sjpadgett) [reply interface/main/tabs/main.php]: Well I was unclear as to how you wanted to use the const so I tested for general values. I'm happy to implement however you need this to be set.
I'd think we'd need a different const if an auto register is desired e.g. OPENEMR_AUTO_REGISTER = {address}.
Maybe best if you just provide a code snippet of how you'd like to see this implemented. I don't have a problem with that at all.

**Comment by sjpadgett** [comment]
Okay. I'll still make your suggested changes but I'm still working on adding the population additions unless this is holding you up.

**Comment by sjpadgett** [comment]
Any suggestions?

```
array (
  'total_patients' => '3',
  'total_portal_patients' => '3',
  'total_users' => '7',
  'users_by_type' => 
  array (
    0 => 
    array (
      'abook_type' => '',
      'user_count' => '6',
    ),
    1 => 
    array (
      'abook_type' => 'external_provider',
      'user_count' => '1',
    ),
  ),
)
```

**Comment by sjpadgett** [comment]
I need opinion on adding portal card click event counts. Working now on modules currently enabled. Any other static data we'd want collected?

**Comment by kojiromike** [comment]
@sjpadgett is there any way we can distinguish via telemetry if the user is using one of the repo docker/*?

**Comment by sjpadgett** [comment]
> repo docker

I'll look into.
I think maybe getenv('OPENEMR_DOCKER_ENV_TAG')
@kojiromike 
We already collect and here is an example snap of partial query from reported

image

**Comment by kojiromike** [comment]
@sjpadgett so if it's docker/production/docker-compose.yml it'd be `production-docker` in telemetry?

**Comment by sjpadgett** [comment]
@kojiromike I suppose. I collect the runtime environment variable  being reported of our dockers. You'd need to point me in a direction if there is additional docker runtime info wanted.

**Comment by bradymiller** [comment]
@kojiromike ,
This is the variable that gets reported (is used on easydev, insanedev, and the demo dockers). Could also place it in the production docker if wish with a tag:
https://github.com/openemr/openemr/blob/master/docker/development-easy/docker-compose.yml#L68

**Comment by bradymiller** [comment]
above link appears to not work (not highlighting the line). it is line 68. the 'OPENEMR_DOCKER_ENV_TAG' setting

**Comment by kojiromike** [comment]
I think it'd be good to know for any of the official openemr deployment strategies, which ones are being used. Maybe each one, and then another one for one that openemr didn't make.

**Comment by sjpadgett** [comment]
I forgot I put portal tracking in so yeah! I could add further tracking but later

image

**Comment by sjpadgett** [comment]
First reporting interval 1203 so telemetry has been out a year in April

Geo tracking only reports for non-usa sites. Errors on IP lookup on US installs

**Comment by sjpadgett** [comment]
Someone needs to look at [Isolated Tests / PHP 8.5 - Isolated Tests (pull_request)]
QueryUtils.php is trying to call the function sqlStatementThrowException(), which is not defined or not available in the current test context.
Location: src/Common/Database/QueryUtils.php:164 Called from: src/Telemetry/TelemetryService.php:145 During test: tests/Tests/Isolated/Telemetry/TelemetryServiceTest.php:432

I don't have a whitespace issue if I run ci and PhpStorm.  I have found that if another test fails for some reason whitespace sometimes fails. Maybe move it up in order.

Still I'm not trying to fix these fails.

File: src/Telemetry/TelemetryRepository.php [hunk 3]
```diff
@@ -67,4 +67,68 @@ public function clearTelemetryData(): void
...
+        // Total active users
+        $sql = "SELECT COUNT(*) AS total_users FROM users WHERE active = 1 AND username IS NOT NULL AND fname IS NOT NULL";
+        $result = $this->fetchRecords($sql, [], noLog: true);
+        $population['total_users'] = $result[0]['total_users'] ?? 0;
+
+        // Active users grouped by abook_type
+        $sql = "SELECT abook_type, COUNT(*) AS user_count 
+            FROM users 
+            WHERE active = 1 AND username IS NOT NULL AND fname IS NOT NULL 
+            GROUP BY abook_type";
```
Comment (kojiromike) [inline src/Telemetry/TelemetryRepository.php]: The trailing whitespace is in this sql. My suggestion would be to write multiline sql using now documents.

```suggestion
        $sql = <<<'SQL'
            SELECT abook_type, COUNT(*) AS user_count
              FROM users
             WHERE active = 1
               AND username IS NOT NULL
               AND fname IS NOT NULL
             GROUP BY abook_type
        SQL;
```

File: src/Telemetry/TelemetryService.php [hunk 4]
```diff
@@ -130,7 +130,6 @@ public function reportUsageData(): int|bool
         $site_uuid = $this->getUniqueInstallationUuid();
         if (empty($site_uuid)) {
             error_log("Site UUID not found.");
-            return false;
```
Comment (kojiromike) [inline src/Telemetry/TelemetryService.php:133]: Removing this return causes the isolated tests to end up in code that they can't run, so they fail.

File: src/Telemetry/TelemetryRepository.php (hunk 3 above)
Comment (sjpadgett) [reply src/Telemetry/TelemetryRepository.php]: I can agree with that. I usually don't do multi line but claude does. I like now document and use often but multi line is used all over core code.
My problem is if my local ide and ci test don't catch, is there any way to give a hint to check for this type of pattern in the fail result?
thanks btw

File: src/Telemetry/TelemetryService.php (hunk 4 above)
Comment (sjpadgett) [reply src/Telemetry/TelemetryService.php:133]: So do I need to fix the test or are you? I don't want to exit here.

Comment (kojiromike) [reply src/Telemetry/TelemetryService.php:133]: Well, the test that's failing is asserting that this function returns `false` when the uuid is not found. So if that's not true anymore, then the test isn't useful anymore.

File: src/Telemetry/TelemetryRepository.php (hunk 3 above)
Comment (kojiromike) [reply src/Telemetry/TelemetryRepository.php]: What IDE do you use?

**Comment by sjpadgett** [comment]
> I don't have a whitespace issue if I run ci and PhpStorm

File: src/Telemetry/TelemetryRepository.php [hunk 5]
```diff
@@ -94,10 +94,13 @@ public function fetchSitePopulationData(): array
         $population['total_users'] = $result[0]['total_users'] ?? 0;
 
         // Active users grouped by abook_type
-        $sql = "SELECT abook_type, COUNT(*) AS user_count 
-            FROM users 
-            WHERE active = 1 AND username IS NOT NULL AND fname IS NOT NULL 
-            GROUP BY abook_type";
+        $sql = <<<SQL
```
Comment (kojiromike) [inline src/Telemetry/TelemetryRepository.php]: If you put single quotes around this it won't allow variables to expand in the text, which means we can use static analysis to reduce the risk of sql injection later.

```suggestion
        $sql = <<<'SQL'
```

**Comment by sjpadgett** [comment]
What do I need to run whitespace local? I finally got my phpstan matching actions run again.

**Comment by kojiromike** [comment]
@sjpadgett install `pre-commit` (https://pre-commit.com/) and set it up to run on your local clone.
